File: src/main/java/org/dxworks/githubminer/dto/export/PullRequestDTO.java
package org.dxworks.githubminer.dto.export;

import com.google.api.client.util.Key;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.dxworks.githubminer.dto.response.repository.pullrequests.PullRequest;

import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class PullRequestDTO {
	@Key
	private Long id;
	@Key
	private Long number;
	@Key
	private String title;
	@Key
	private String body;
	@Key
	private PrBranchDTO head;
	@Key
	private PrBranchDTO base;
	@Key
	private List<String> commits;
	@Key
	private String createdAt;
	@Key
	private String updatedAt;
	@Key
	private String mergedAt;
	@Key
	private String closedAt;
	@Key
	private String state;
	@Key
	private UserDTO createdBy;
	@Key
	private UserDTO assignee;
	@Key
	private UserDTO mergedBy;
	@Key
	private List<PullRequestCommentDTO> comments;
	@Key
	private List<PullRequestReviewDTO> reviews;

	public static PullRequestDTO fromPullRequest(PullRequest pullRequest) {
		if (pullRequest == null || com.google.api.client.util.Data.isNull(pullRequest))
			return null;
		return builder()
				.id(pullRequest.getId())
				.number(pullRequest.getNumber())
				.title(pullRequest.getTitle())
				.body(pullRequest.getBody())
				.head(PrBranchDTO.fromPullRequestBranch(pullRequest.getHead()))
				.base(PrBranchDTO.fromPullRequestBranch(pullRequest.getBase()))
				.createdAt(pullRequest.getCreatedAt())
				.mergedAt(pullRequest.getMergedAt())
				.updatedAt(pullRequest.getUpdatedAt())
				.closedAt(pullRequest.getClosedAt())
				.state(pullRequest.getState())
				.createdBy(UserDTO.fromUser(pullRequest.getUser()))
				.assignee(UserDTO.fromUser(pullRequest.getAssignee()))
				.mergedBy(UserDTO.fromUser(pullRequest.getMergedBy()))
				.build();
	}
}

File: src/main/kotlin/org/dxworks/githubminer/dto/export/PullRequestDTO.kt
package org.dxworks.githubminer.dto.export

import com.google.api.client.util.Data
import com.google.api.client.util.Key
import org.dxworks.githubminer.dto.response.repository.pullrequests.PullRequest

class PullRequestDTO (
    @Key
    var id: Long? = null,

    @Key
    var number: Long,

    @Key
    var title: String? = null,

    @Key
    var body: String? = null,

    @Key
    var head: PrBranchDTO? = null,

    @Key
    var base: PrBranchDTO? = null,

    @Key
    var commits: List<String>? = null,

    @Key
    var createdAt: String? = null,

    @Key
    var updatedAt: String? = null,

    @Key
    var mergedAt: String? = null,

    @Key
    var closedAt: String? = null,

    @Key
    var state: String? = null,

    @Key
    var createdBy: UserDTO? = null,

    @Key
    var assignee: UserDTO? = null,

    @Key
    var mergedBy: UserDTO? = null,

    @Key
    var comments: List<PullRequestCommentDTO>? = null,

    @Key
    var reviews: List<PullRequestReviewDTO>? = null
) {
    companion object {
        fun fromPullRequest(pullRequest: PullRequest?): PullRequestDTO? {
            return if (pullRequest == null || Data.isNull(pullRequest)) null else PullRequestDTO(
                    id = pullRequest.id,
                    number = pullRequest.number,
                    title = pullRequest.title,
                    body = pullRequest.body,
                    head = PrBranchDTO.fromPullRequestBranch(pullRequest.head),
                    base = PrBranchDTO.fromPullRequestBranch(pullRequest.base),
                    createdAt = pullRequest.createdAt,
                    mergedAt = pullRequest.mergedAt,
                    updatedAt = pullRequest.updatedAt,
                    closedAt = pullRequest.closedAt,
                    state = pullRequest.state,
                    createdBy = UserDTO.fromUser(pullRequest.user),
                    assignee = UserDTO.fromUser(pullRequest.assignee),
                    mergedBy = UserDTO.fromUser(pullRequest.mergedBy)
            )
        }
    }
}
